src/server/game/Entities/AreaTrigger/AreaTrigger.cpp

```diff
@@ -642,6 +659,60 @@ bool AreaTrigger::HasSplines() const
     return bool(_spline);
 }
 
+void AreaTrigger::InitCircularMovement(AreaTriggerCircularMovementInfo const& cmi, uint32 timeToTarget)
+{
+    // should be sent in object create packets only
+    m_uint32Values[AREATRIGGER_TIME_TO_TARGET] = timeToTarget;
+
+    _areaTriggerCircularMovementInfo = cmi;
+
+    _areaTriggerCircularMovementInfo.TimeToTarget = GetTimeToTarget();
+    _areaTriggerCircularMovementInfo.ElapsedTimeForMovement = GetElapsedTimeForMovement();
+
+    if (IsInWorld())
+    {
+        // TODO: broadcast AreaTriggerReShape?
+    }
+}
+
+Position const* AreaTrigger::GetCircularMovementCenterPosition() const
+{
+    if (!_areaTriggerCircularMovementInfo.Center.is_initialized() && _areaTriggerCircularMovementInfo.TargetGUID.is_initialized())
+        if (WorldObject* center = ObjectAccessor::GetWorldObject(*this, *_areaTriggerCircularMovementInfo.TargetGUID))
+            return center;
+
+    return &_areaTriggerCircularMovementInfo.Center->Pos;
+}
+
+void AreaTrigger::UpdateCircularMovementPosition()
+{
+    Position const* centerPos = GetCircularMovementCenterPosition();
+    if (!centerPos)
+        return;
+
+    AreaTriggerCircularMovementInfo const& cmi = GetMiscTemplate()->CircularMovementInfo;
+
+    // AreaTrigger make exactly "TimeToTarget / Duration" loops during his life time
+    float angleDiff = 2.0f * float(M_PI) * (float(GetTimeSinceCreated()) / float(GetTimeToTarget()));
+
+    // Adapt angle diff depending of circle direction
+    angleDiff *= cmi.CounterClockwise ? 1 : -1;
+
+    // We already made one circle & can't loop
+    if (!cmi.CanLoop && (angleDiff <= 0.0f || angleDiff >= 2.0f * float(M_PI)))
```

Comment:
can that work clockwise?

Reply:
You mean if this calculation can work clockwise?

Reply:
yes? angleDiff is always < 0

Reply:
Yeah well.. actually it doesnt even matter if clockwise or not.. if CanLoop == false.. then the whole movement doesnt work :D